scripts/word_embeddings/tools/vocab_count.cc

```diff
@@ -0,0 +1,151 @@
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <numeric>
+#include <queue>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "CLI/CLI.hpp"    // command line parser
+#include "sparsepp/spp.h" // fast sparse hash map
+
+#include "./utils.h"
+
+namespace fs = std::filesystem;
+using Vocab = spp::sparse_hash_map<std::string, uint32_t>;
+
+std::mutex paths_m;
+std::mutex vocabs_m;
+
+void readVocab(std::queue<fs::path> &paths, queue<Vocab> &vocabs) {
+  std::unique_ptr<Vocab> vocab = std::make_unique<Vocab>();
+  std::string word;
+  fs::path path;
+
+  while (true) {
+    {
+      std::scoped_lock lock(paths_m);
+      if (paths.empty()) {
+        break;
+      }
+      path = paths.front();
+      paths.pop();
+    }
+
+    std::ifstream in{path};
+    if (!in.is_open()) {
+      throw std::invalid_argument(path.string() + " cannot be opened!");
+    }
+    while (in >> word) {
+      (*vocab)[word]++;
+    }
+  }
+  {
+    std::scoped_lock lock(vocabs_m);
+    vocabs.push(std::move(vocab));
+  }
+}
+
+std::unique_ptr<Vocab> combineVocabs(queue<Vocab> &vocabs, int numThreads) {
+  std::unique_ptr<Vocab> vocab1 = vocabs.pop();
+  for (int i = 1; i < numThreads; i++) {
+    std::unique_ptr<Vocab> vocab2 = vocabs.pop();
+    if (vocab1->size() < vocab2->size()) {
+      for (const auto &e : *vocab1) {
+        (*vocab2)[e.first] += e.second;
+      }
+      std::swap(vocab1, vocab2);
+    } else {
+      for (const auto &e : *vocab2) {
+        (*vocab1)[e.first] += e.second;
+      }
+    }
+  }
+  return vocab1;
+}
+
+int main(int argc, char **argv) {
+  // Performance optimizations for writing to stdout
+  std::ios::sync_with_stdio(false);
+
+  CLI::App app("Simple tool to extract unigram counts");
+  std::vector<std::string> files;
+  app.add_option("FILES", files, "File names")->check(CLI::ExistingPath);
+  unsigned int minCount = 5;
+  app.add_option("-c,--minCount", minCount,
+                 "Minimum number of occurences required for a word to be "
+                 "included in the vocabulary.");
+  unsigned int numThreads = 1;
+  app.add_option(
+         "-j,--numThreads", numThreads,
+         "Number of threads to use. Each thread constructs an "
+         "independent vocabulary which are finally merged. Only appropriate "
+         "when multiple, sufficiently large input files are specified.")
+      ->check(CLI::Range(1U, std::numeric_limits<unsigned int>::max()));
+  CLI11_PARSE(app, argc, argv);
+
+  std::queue<fs::path> paths;
+  for (auto &file : files) {
+    paths.emplace(file);
+  }
+  std::vector<std::thread> threads;
+  queue<Vocab> vocabs;
+  for (unsigned int i = 0; i < numThreads; i++) {
+    threads.push_back(std::thread(
+        [&paths, &vocabs]() { readVocab(std::ref(paths), std::ref(vocabs)); }));
+  }
+  std::unique_ptr<Vocab> vocab = combineVocabs(vocabs, numThreads);
```

Comment:
It is required that `combineVocabs` starts after all `readVocab` threads are done? Is that enforced by the blocking queue?

Reply:
Yes, enforced by blocking Queue. This is better than starting combineVocabs after all readVocabs are done, because the combination can run in parallel with the reading.